docs/proposals/horovod-on-tony.md

```diff
@@ -0,0 +1,113 @@
+
+## API
+There are no API changes to other machine learning frameworks.
+
+## Design
+Attention: It's gloo controller that makes TonY support horovod.
+
+From the perspective of compatibility and maintainability, it's better to directly use the [gloo_runner](https://github.com/horovod/horovod/blob/master/horovod/runner/gloo_run.py) on TonY. But after reading through Horovod's code, I find it difficult to reuse gloo_runner's code on TonY, because the existence of some unrelated codes will lead driver to start the worker through the SSH command, which is hard to be supportted on Yarn.
+
+After having a deep understanding of Horovod code and [communicating with developers](https://github.com/horovod/horovod/discussions/2785), i know that the Gloo controller uses a rendezvous server to assign each worker role, and provides HTTP API for workers to obtain cluster information. So each worker  can build a training cluster and start training at the same time.
+
+Horovod is served as two roles, worker and driver. Driver is responsible for starting the rendezvous server and will not participate in training (no GPU required, lightweight). Before starting, driver need to know all workers' hostnames in advance. The worker is only responsible for training. According to TonY's architecture (**Application master** and **task executor**), the design can be as follows.
+
+### Horovod Driver
+__How to start rendezvous server__  
+Reusing Horovod rendezvous server code, we introduce tony-horovod driver launcher to offer a python script
+```python
+# Init the horovod rendezous server
+global_rendezv = RendezvousServer(verbose=1)
+# Output server port, which will be used horovod worker to connect server.
+global_rendezv_port = global_rendezv.start()
+print("Rendezvous server started, port: " + str(global_rendezv_port))
+
+hosts = parse_hosts(worker_list)
+# Output the host plan, it will output local_rank, rank and so on.
+host_alloc_plan = get_host_assignments(hosts, 1)
+
+# Start the server.
+global_rendezv.init(host_alloc_plan)
+```
+
+__When to start driver__  
+After all workers' resource have be assigned and TonY's Application master could get all workers' registry info.
+
+__Where to start driver__  
+Two options  
+1. On TonY application master.  
+This will save resources(no extra resources to start driver), and the amount of code changes will be small. But by injecting relevant Horovod's driver code into AM, it is not elegant.
+2. On TonY task executor.  
+Additional customization of the driver configuration is required and allow different task executors to specify different resources(because driver needn't gpu). And it is necessary to coordinate the startup sequence between the driver and other workers, because driver should start before worker. But 
+we should make some changes to prevent starting in each task executor at the same time.  
+
+First option will be adopted. The second one remains to be supported in the future.
```

Comment:
I prefer using #2 . 
- AM should only be used for container orchestration. We don't want any framework specific code to run there.
- TonY supports a mini DAG engine: https://github.com/linkedin/TonY/blob/master/tony-core/src/main/java/com/linkedin/tony/TonyConfigurationKeys.java#L131 , tasks in prepare-stage will finish first before training-stage starts.

Reply:
Good idea. It will work by using **prepare-stage**.

Reply:
It's hard to reuse prepare-stage conf, TonY will track and wait for to finish before starting tasks in the training stage.
However, we just need to keep startup sequence and ensure that the **driver** and **worker** life cycles are the same.

Reply:
@UWFrankGu  how did we use `pre-stage` to load data into redis before starting the training job?

Reply:
We added `redis` servers in `tony.application.untracked.jobtypes` to ensure that the `redis` process runs throughout `prepare-stage` and `training-stage`. More information in the TonY Configurations doc: https://github.com/linkedin/TonY/wiki/TonY-Configurations

Reply:
ah right.. you also need `tony.application.untracked.jobtypes` 😂

Reply:
I think i will not use `stage` to coordinate driver and worker startup sequence.
Through extending the `canStart()` method in `MLFrameworkRuntime` interface, it's more elegant to coordinate each task executor's python process startup sequence.